File: grpc/hello-grpc-csharp/HelloClient/ProtoClient.cs
using System;
using System.IO;
using System.Reflection;
using System.Threading.Tasks;
using Common;
using Grpc.Core;
using log4net;
using log4net.Config;
using Org.Feuyeux.Grpc;

namespace HelloClient
{
    public class ProtoClient
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(ProtoClient));
        private readonly LandingService.LandingServiceClient _client;

        private ProtoClient(LandingService.LandingServiceClient client)
        {
            this._client = client;
        }

        public static void Main()
        {
            var logRepository = LogManager.GetRepository(Assembly.GetEntryAssembly());
            XmlConfigurator.Configure(logRepository, new FileInfo("log4net.config"));

            var channel = Connection.GetChannel();
            var client = new ProtoClient(new LandingService.LandingServiceClient(channel));
            Log.Info("Unary RPC");
            client.Talk();
            Log.Info("Server streaming RPC");
            client.TalkOneAnswerMore().Wait();
            Log.Info("Client streaming RPC");
            client.TalkMoreAnswerOne().Wait();
            Log.Info("Bidirectional streaming RPC");
            client.TalkBidirectional().Wait();
            channel.ShutdownAsync().Wait();
            Log.Info("DONE");
        }

        private void Talk()
        {
            var request = new TalkRequest
            {
                Data = "0",
                Meta = "C#"
            };
            try
            {
                Log.Info($"Request: data={request.Data},meta={request.Meta}");
                var talkResponse = _client.Talk(request, BuildHeaders());
                PrintResponse(talkResponse);
            }
            catch (RpcException e)
            {
                Console.Error.WriteLine(e);
            }
        }

        private async Task TalkOneAnswerMore()
        {
            try
            {
                var request = new TalkRequest
                {
                    Data = "0,1,2",
                    Meta = "C#"
                };
                Log.Info($"Request: data={request.Data},meta={request.Meta}");
                using var call = _client.TalkOneAnswerMore(request, BuildHeaders());
                var responseStream = call.ResponseStream;
                while (await responseStream.MoveNext())
                {
                    var talkResponse = responseStream.Current;
                    PrintResponse(talkResponse);
                }
            }
            catch (RpcException e)
            {
                Console.Error.WriteLine(e);
            }
        }

        private async Task TalkMoreAnswerOne()
        {
            try
            {
                using var call = _client.TalkMoreAnswerOne(BuildHeaders());
                var requests = Utils.BuildLinkRequests();
                foreach (var request in requests){
                    Log.Info($"Request: data={request.Data},meta={request.Meta}");
                    await call.RequestStream.WriteAsync(request);
                    await Task.Delay(Utils.HelloRandom.Next(100) + 100);
                }
                await call.RequestStream.CompleteAsync();
                var talkResponse = await call.ResponseAsync;
                PrintResponse(talkResponse);
            }
            catch (RpcException e)
            {
                Console.Error.WriteLine(e);
            }
        }

        private async Task TalkBidirectional()
        {
            try
            {
                using var call = _client.TalkBidirectional(BuildHeaders());
                var responseReaderTask = Task.Run(async () =>
                {
                    while (await call.ResponseStream.MoveNext())
                    {
                        var talkResponse = call.ResponseStream.Current;
                        PrintResponse(talkResponse);
                    }
                });
                var requests = Utils.BuildLinkRequests();
                foreach (var request in requests)
                {
                    Log.Info($"Request: data={request.Data},meta={request.Meta}");
                    await call.RequestStream.WriteAsync(request);
                    await Task.Delay(Utils.HelloRandom.Next(100) + 100);
                }
                await call.RequestStream.CompleteAsync();
                await responseReaderTask;
            }
            catch (RpcException e)
            {
                Console.Error.WriteLine(e);
            }
        }

        private static Metadata BuildHeaders()
        {
            var headers = new Metadata
            {
                new Metadata.Entry("k1", "v1"),
                new Metadata.Entry("k2", "v2")
            };
            return headers;
        }

        private static void PrintResponse(TalkResponse talkResponse)
        {
            foreach (var result in talkResponse.Results)
            {
                var kv = result.Kv;
                Log.Info($"{talkResponse.Status} {result.Id} [{kv["meta"]} {result.Type} {kv["id"]},{kv["idx"]}:{kv["data"]}]");
            }
        }
    }
}

File: hello-grpc-cpp-bazel/client/proto_client.cpp
#include <iostream>
#include <memory>
#include <string>

#include "grpcpp/grpcpp.h"
#include "glog/logging.h"
#include <thread>

#include "protos/landing.grpc.pb.h"
#include "common/utils.h"
#include "common/connection.h"

using google::protobuf::Map;
using google::protobuf::RepeatedPtrField;
using grpc::Channel;
using grpc::ClientContext;
using grpc::ClientReader;
using grpc::ClientReaderWriter;
using grpc::ClientWriter;
using grpc::Status;
using hello::Connection;
using hello::Utils;
using org::feuyeux::grpc::LandingService;
using org::feuyeux::grpc::ResultType;
using org::feuyeux::grpc::TalkRequest;
using org::feuyeux::grpc::TalkResponse;
using org::feuyeux::grpc::TalkResult;
using std::string;

class LandingClient
{
public:
    explicit LandingClient(const std::shared_ptr<Channel> &channel) : client(LandingService::NewStub(channel)) {}

    void Talk()
    {
        ClientContext context;
        buildHeaders(context);
        TalkResponse talkResponse;
        TalkRequest talkRequest;
        talkRequest.set_data("1");
        talkRequest.set_meta("C++");
        Status status = client->Talk(&context, talkRequest, &talkResponse);
        if (status.ok())
        {
            printResponse(context, talkResponse);
        }
        else
        {
            LOG(INFO) << "Error:" << status.error_code() << ": " << status.error_message();
        }
    }

    void TalkOneAnswerMore()
    {
        ClientContext context;
        buildHeaders(context);
        TalkResponse talkResponse;
        TalkRequest talkRequest;
        talkRequest.set_data("1,2,3");
        talkRequest.set_meta("C++");
        const std::unique_ptr<::grpc::ClientReader<TalkResponse>>
            &response(client->TalkOneAnswerMore(&context, talkRequest));
        while (response->Read(&talkResponse))
        {
            printResponse(context, talkResponse);
        }
    }

    void TalkMoreAnswerOne()
    {
        ClientContext context;
        buildHeaders(context);
        TalkResponse talkResponse;
        std::unique_ptr<ClientWriter<TalkRequest>> writer(client->TalkMoreAnswerOne(&context, &talkResponse));
        const list<TalkRequest> &requests = Utils::buildLinkRequests();
        for (auto &&request : requests)
        {
            if (!writer->Write(request))
            {
                // Broken stream.
                break;
            }
            std::this_thread::sleep_for(std::chrono::milliseconds(200));
        }
        writer->WritesDone();
        Status status = writer->Finish();
        if (status.ok())
        {
            printResponse(context, talkResponse);
        }
        else
        {
            LOG(INFO) << "Error:" << status.error_code() << ": " << status.error_message();
        }
    }

    void TalkBidirectional()
    {
        ClientContext context;
        buildHeaders(context);
        TalkResponse talkResponse;
        std::shared_ptr<ClientReaderWriter<TalkRequest, TalkResponse>> stream(client->TalkBidirectional(&context));
        std::thread writer([stream]()
                           {
            const list<TalkRequest> &requests = Utils::buildLinkRequests();
            for (auto &&request: requests) {
                stream->Write(request);
            }
            stream->WritesDone(); });
        while (stream->Read(&talkResponse))
        {
            printResponse(context, talkResponse);
        }
        writer.join();
        Status status = stream->Finish();
        if (!status.ok())
        {
            LOG(INFO) << "Error:" << status.error_code() << ": " << status.error_message();
        }
    }

    static void printResponse(ClientContext &context, const TalkResponse &response)
    {
        const multimap<grpc::string_ref, grpc::string_ref> &headers = context.GetServerInitialMetadata();
        for (const auto &header : headers)
        {
            LOG(INFO) << "<-H " << header.first << ":" << header.second;
        }
        const RepeatedPtrField<TalkResult> &talkResults = response.results();
        for (const TalkResult &result : talkResults)
        {
            const Map<string, string> &kv = result.kv();
            string id(kv.at("id"));
            string idx(kv.at("idx"));
            string meta(kv.at("meta"));
            string data(kv.at("data"));
            LOG(INFO) << response.status()
                      << " " << result.id()
                      << " [" << meta
                      << " " << ResultType_Name(result.type())
                      << " " << id
                      << " " << idx
                      << " " << data
                      << "]";
        }
        const multimap<grpc::string_ref, grpc::string_ref> &tails = context.GetServerTrailingMetadata();
        for (const auto &tail : tails)
        {
            LOG(INFO) << "<-L " << tail.first << ":" << tail.second;
        }
    }

    static void buildHeaders(ClientContext &context)
    {
        // examples/cpp/metadata
        context.AddMetadata("k1", "v1");
        context.AddMetadata("k2", "v2");
    }

private:
    std::unique_ptr<LandingService::Stub> client;
};

int main(__attribute__((unused)) int argc, char **argv)
{
    Utils::initLog(argv);
    LandingClient landingClient(Connection::getChannel());
    LOG(INFO) << "Unary RPC";
    landingClient.Talk();
    LOG(INFO) << "Server streaming RPC";
    landingClient.TalkOneAnswerMore();
    LOG(INFO) << "Client streaming RPC";
    landingClient.TalkMoreAnswerOne();
    LOG(INFO) << "Bidirectional streaming RPC";
    landingClient.TalkBidirectional();
    google::ShutdownGoogleLogging();
    return 0;
}